[IDEA] Adding the detailed version of the muon system.

mahmoudali2 opened the pull request:
BEGINRELEASENOTES
- Defining the first draft of the detailed muon system, which depend on mosaics of 50 * 50 cm^2 mRWELL chambers.
- Defining a suitable XML for the new detailed version.
- Describing µRWELL materials.
- Adding the parameters of the muon system into the full IDEA implementation.
- ...

ENDRELEASENOTES

BrieucF commented on detector/muonSystem/README.md:
I propose to remove references to IDEA in this README, the detector builder is quite generic and could be used for other implementation. e.g. `IDEA Muon System` --> `µRWELL Based Muon System`

mahmoudali2 replied:
Hi Brieuc, I have edited the code after checking all the overlaps, and now it's clean.. I'm attaching the overlapDump file. 
[overlapDump-151.txt](https://github.com/key4hep/k4geo/files/15309202/overlapDump-151.txt)

BrieucF commented on detector/muonSystem/muonSystem_v00.cpp:
Can you change the name of the driver to `muonSystemMuRWELL_o1_v01` or equivalent?

BrieucF commented on detector/muonSystem/README.md:
```diff
@@ -0,0 +1,11 @@
+The code has been designed to very flexible with the modifications, where the user can choose the number of sides of the shape `numSides` (hexagon, octagon, ....etc), and automatically the builder will calculate the number and places of the copied chambers. Some of the code advantages: 
+ * If the side length do not fit with an integer number of 50 × 50 cm² , the builder can make a chamber with unusual dimensions, which can fit the excess area at the end of the side.
+ * The availability to make multiple layers with different inner radius and barrel length.
+ * The code is very general, which can be used to describe any detector system made from repeated tiles, with having the capability to fill the gaps with unusual dimensions tiles.
```
```suggestion
 * The code is very general, it can be used to describe any detector system made from repeated tiles (e.g. pre-shower) and has the capability to fill the gaps with unusual dimensions tiles.
```

BrieucF commented on FCCee/IDEA/compact/IDEA_o1_v02/detailedMuonSystem_v00.xml:
rename to `(...)o1_v01.xml`

BrieucF commented on detector/muonSystem/muonSystem_v00.cpp:
```diff
@@ -0,0 +1,891 @@
+}
+}
+
+DECLARE_DETELEMENT(muonSystem_o1_v00, det::createmuonSystem_o1_v00)
```
Let make sure the name of the DETELEMENT follows the name of the file

BrieucF commented on FCCee/IDEA/compact/IDEA_o1_v02/detailedMuonSystem_v00.xml:
```diff
@@ -0,0 +1,210 @@
+      <dimensions x="mRWELLTotalThickness/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="0*mm" y_offset="0*mm" z_offset="0*mm" material="Air"/>
+      <sensitive type="SimpleTrackerSD"/>
+
+      <!-- Specify the detector paramenters and the overlap  // radius is put in the middle, so its not the inner neither the outer // the endcap detector layers z-offset values have to be changed.. now the layer is thicker = 12 cm -->
```
```suggestion
      <!-- Specify the detector parameters and the overlap  // radius is put in the middle, so its not the inner neither the outer // the endcap detector layers z-offset values have to be changed.. now the layer is thicker = 12 cm -->
```

BrieucF commented on FCCee/IDEA/compact/IDEA_o1_v02/detailedMuonSystem_v00.xml:
```diff
@@ -0,0 +1,210 @@
+  <detectors>
+
+    <!-- mRWELL envelope -->
+    <detector name="muonSystem1" type="muonSystem_o1_v00"  vis="MuonVisEnv" id="51" readout="mRWELLChamberReadout">
```
Here and in other places: the "id" should be defined in `DectDimensions_IDEA_o1_v01.xml` and called here (see and modify https://github.com/key4hep/k4geo/blob/a02cbb463f412055d0c788f58b799e0b95cc8aae/FCCee/IDEA/compact/IDEA_o1_v02/DectDimensions_IDEA_o1_v01.xml#L43)

BrieucF replied:
I also suspect this should crash as it is now (did you try running it with ddsim?) because we have 4 bits for `system` in the readout which only allows to have "id's" from 0 to 15.

mahmoudali2 replied:
I added the 9 muon system layers IDs to the DectDimensions_IDEA_o1_v01.xml, and ddsim run without crashes even if the system has 4 bits. Isn't the 4 bits in that case only for the muon system, since for example I have here 9 detectors (layers)? since also every sub-detector has its own readout description?

BrieucF commented on detector/muonSystem/muonSystem_v00.cpp:
```diff
@@ -0,0 +1,891 @@
+Factory for IDEA muon system
+Expected xml structure (the 'sensitive' keyword is optional and defaults to false):
+<detector type="muonSystem_o1_v00" ...>
+  <dimensions x="..." y="..." z="..." z_offset="..." x_offset="..." y_offset="...">
```
can you comment here what the various parameters (x, y, z, ...) mean exactly?

BrieucF commented on detector/muonSystem/muonSystem_v00.cpp:
```diff
@@ -0,0 +1,891 @@
+  <dimensions x="..." y="..." z="..." z_offset="..." x_offset="..." y_offset="...">
+  <sensitive type="SimpleTrackerSD"/>
+
+  <!-- Specify the detector paramenters and the overlap // if you want exclude any component (e.g: endcap, radiators ..), just but its num =0 // radius is put in the middle, so its not the inner neither the outer  -->
```
```suggestion
  <!-- Specify the detector paramenters and the overlap // if you want exclude any component, e.g: endcap, just put endcapDetectorParameters=0 // radius is put in the middle, so its not the inner neither the outer  -->
```?

BrieucF reviewed (commented):
Hi Mahmoud, thanks a lot for this! Here are already a few comments. Can you also try to run the overlap check?
```
create a geant4 macro named overlap.mac with this inside:
/geometry/test/run
exit

Then run the overlap checker with this: 
ddsim --compactFile ./compact/simple_detector.xml --runType run --part.userParticleHandler='' --macroFile overlap.mac >> overlapDump.txt
```

mahmoudali2 commented:
New modification in the builder and the compact file to avoid all the overlap cases, I have introduced a new way to organize the different sides in the same layer, the following photo (X-Y plane) is an example for barrel sides overlapping and avoiding touching each others:
![barrel-side-overlap](https://github.com/key4hep/k4geo/assets/120766336/cc48d85c-23b8-4104-81a0-5f2a26a9ab8a)
The same in encap, with creating unusual shapes by boolean operations `UnionSolid` (The only solution I got to solve the overlap with the mother volume): 
![endcap-overlapping](https://github.com/key4hep/k4geo/assets/120766336/8524494d-209b-4f50-9792-5407960e5f5f)
...
And every layer (1 detector layer + 1 yoke layer "if present") has an independent mother volume in the shape of `Polyhedra` to avoid overlapping with each other: 
![layer](https://github.com/key4hep/k4geo/assets/120766336/d5e61783-1361-4dbb-8d83-c4fc427bf2e4)
....
Overall the complete muon system: 
![muon](https://github.com/key4hep/k4geo/assets/120766336/89075863-77fd-4b2b-a3c0-5163d5a39cad)

BrieucF reviewed (commented):
Hi Mahmoud, this is fantastic! Tahnks! Here is a first batch of comment. 
Can you also update `FCCee/IDEA/compact/README.md` under IDEA_o1_v02 with something like:
"Based on o1_v01 but with a detailed description of the vertex detector, drift chamber and muon system, a place holder solenoid and the endplate absorber. Missing: SiWrapper, calorimeter."?

I am also a bit puzzled by the fact of having one detID per layer but I guess having one detID for the whole barrel and one for the whole endcap would imply a complete refactoring of the code?

mahmoudali2 commented on FCCee/IDEA/compact/IDEA_o1_v03/detailedMuonSystem_o1_v01.xml:
```diff
@@ -0,0 +1,321 @@
+      <endcapDetectorParameters numEndcapDetectorLayers="0" endcapDetectorVolumeThickness="microRWELLVolumeThickEndcap" endcapDetectorLayerInnerRadius="EndcapLayersInnerRadius" endcapDetectorLayerOuterRadius="EndcapLayersOuterRadius" endcapDetectorZOffset="EndcapFirstLayerZOffset+microRWELLVolumeThickEndcap/2.0" />
+
+      <!-- mRWELL chamber -->
+      <layer x="G10_FR4Thick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick/2.0" y_offset="0*mm" z_offset="0*mm" material="G10_FR40x5633439ca570" vis="MuonVis" />
+      <layer x="CuThick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick/2.0" y_offset="0*mm" z_offset="0*mm" material="G4_Cu"  vis="MuonVis" />
+      <layer x="GasLayerThick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick+GasLayerThick/2.0" y_offset="0*mm" z_offset="0*mm" material="ARCO2CF4" sensitive="true" vis="MuonVis" />
+      <layer x="Cu2Thick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick/2.0" y_offset="0*mm" z_offset="0*mm" material="G4_Cu"  vis="MuonVis" />
+      <layer x="KaptonThick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick+KaptonThick/2.0" y_offset="0*mm" z_offset="0*mm" material="Kapton" vis="MuonVis" />
+      <layer x="CarnonFiberThick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick+KaptonThick+CarnonFiberThick/2.0" y_offset="0*mm" z_offset="0*mm" material="CarbonFiber" vis="MuonVis" />
+      <layer x="CuThick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick+KaptonThick+CarnonFiberThick+CuThick/2.0" y_offset="0*mm" z_offset="0*mm" material="G4_Cu"  vis="MuonVis" />
+      <layer x="CarbonFiber2Thick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick+KaptonThick+CarnonFiberThick+CuThick+CarbonFiber2Thick/2.0" y_offset="0*mm" z_offset="0*mm" material="CarbonFiber" vis="MuonVis" />
+      <layer x="CuThick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick+KaptonThick+CarnonFiberThick+CuThick+CarbonFiber2Thick+CuThick/2.0" y_offset="0*mm" z_offset="0*mm" material="G4_Cu"  vis="MuonVis" />
+      <layer x="SiThick/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="-mRWELLTotalThickness/2.0+G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick+KaptonThick+CarnonFiberThick+CuThick+CarbonFiber2Thick+CuThick+SiThick/2.0" y_offset="0*mm" z_offset="0*mm" material="G4_Si0x56335d3d1480" vis="MuonVis" />
+ 
```
I plan actually to do that, a briefer way to describe the muRWELL chamber in the xml. But, I propose to let it be in the next version, since it requires some changes in the builder.

BrieucF commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,941 @@
+Expected xml structure (the 'sensitive' keyword is optional and defaults to false):
+<detector type="muonSystemMuRWELL_o1_v01" ...>
+  <envelope rmin="..." rmax="..." z="..."  z_offset="..." material="..."/>  <!-- dimensions of the whole detector envelope-->
+  <dimensions x="..." y="..." z="..." z_offset="..." x_offset="..." y_offset="...">  <!--  dimension of the local chamber envelope. x: the thickness of the chamber, y&z: the 2D plane dimensions of the chambers-->
```
They seem to be half-lengths? Is it also true for the x-coordinate? I think it worth mentioning this.

BrieucF commented on FCCee/IDEA/compact/IDEA_o1_v03/DectDimensions_IDEA_o1_v02.xml:
```diff
@@ -220,6 +230,33 @@
+    <constant name = "numberOfSides"                              value = "8"/>        <!-- The number of sides of the muon system e.g (Octagon, Hexagon, ...)--> 
+
+                                                <!-- Barrel -->
+    <constant name = "BarrelFirstLayerRadius"                     value = "4500*mm"/>  <!-- 1st Barrel microRWELL detector inner radius-> its the start point of thicknesses of the microRWELL material -->
```
at the middle of the "side"?

andresailer commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,941 @@
+        
+      for (int rectangle = 0; rectangle < (numRectangles + 1); ++rectangle) {
+
+          double rectangleEnvX = detectorVolumeThickness/4.5;  // dividing by 4.5 gets the best thickness for the recatngle to avoid any overlap ~ in our case the uRWELL the best rectangle thickness is 26.667 mm which is 120/4.5.
```
This  `/4.5` seems awfully fine-tuned? Or is this OK also for other thinknesses?

mahmoudali2 replied:
Hi @andresailer, I have changed the parameters, and now it is OK with other thicknesses to. Whatever the chamber dimensions the user will put in the compact file, It will work. I have made the variable  `detectorVolumeThickness` is determined by the builder depending on the chamber dimensions and the overlap angle, instead before it was determined by the user.

andresailer commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,941 @@
+//                               // B A R R E L //
+// ----------------------------------------------------------------------------------------------------
+
+  if (numBarrelDetectorLayers > 0){
```
Maybe it makes sense to split this into a barrel and an endcap constructor?

mahmoudali2 replied:
Since the barrel last layer is enclosing all the endcap layers inside, we preferred to keep them in the same builder. Both Barrel and Endcap are sharing some common variables like `numSides` of the main polygon shape. and the length of the last layer of the barrel depending on the total thickness of the endcap detector. And to distinguish between the collection hits of barrel and endcap, a bit-field component has been defined. Maybe we might tend to separate them in the future if needed.

andresailer commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,941 @@
+
+  int numBarrelLayer = 1;
+  double barrelLayerRMin = radius - detectorVolumeThickness/2.0;   // have to be changed depends on xml impementation
+  double barrelLayerRMax = radius + 1.5 * detectorVolumeThickness;
```
Is this 1.5 factor always correct?

mahmoudali2 replied:
Yes, since the detector volume was centered in the 0.25 of the thickness of that volume. But I have also changed it in the new commits, and I centered it in the middle instead.

andresailer commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,941 @@
+  double angle_clearance = 0.0; // 0.02 works good but needs the detector volume thick to be more than 60 mm. // it's less than 1 degree, we use the clearnce to avoid overlapping
+
+  double sideLength = 2 * radius * std::tan(shapeAngle_radians); 
+  double sideLength2 = 2 * (radius + 1.5 * detectorVolumeThickness) * std::tan(shapeAngle_radians);
```
What does the 2 signify?

mahmoudali2 replied:
It's coming from the formula to calculate the polygon side length `Side Length a. a = 2r tan(π/n)`.

mahmoudali2 replied:
If you mean the number 2 in `sideLength2 ` and other variable as well, that represents here the trapezoid's **parallel sides** lengths. just to distinguish between the two sides lengths.

andresailer replied:
Sorry, yes I mean the 2 in `sideLength2`, can you please add what you wrote as a comment.

mahmoudali2 replied:
Done, I Have added the comments.

andresailer commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,941 @@
+      dd4hep::PlacedVolume sidePhys;
+      dd4hep::DetElement sideDE;
+
+      if (side % 2 == 0) {
```
It feels like this can be much simplified. Many calculations are done and then never used depending on `side % 2`

andresailer commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,941 @@
+
+        //    int barrelChamberID = side * 1000 + rectangle * 100 + chamberIndex;  // later you should add layer number to distinguish between different barrel layers.
+            
+            std::stringstream barrelNameStream;
```
indentation is off

BrieucF commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,928 @@
+  double endcapDetectorSideBoxYLength = 2 * dimensions.z();
+
+  double endcapRadiatorSideLength = 2 * (endcapRadiatorLayerInnerRadius) * std::tan(shapeAngle_radians); //it is also the same for the endcap detector layers.
+  double endcapRadiatorSideLength2 = 2 * (endcapRadiatorLayerOuterRadius) * std::tan(shapeAngle_radians); //it is also the same for the endcap detector layers.
```
can you add a comment here, saying they are the two sides of the trapezoid?

BrieucF commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,928 @@
+  dd4hep::Volume BarrelDetectorLayerVolume(barrelDetectorName, BarrelDetectorLayer, mat);
+
+  double sideLength = 2 * (barrelLayerRMid - detectorVolumeThickness/2.0) * std::tan(shapeAngle_radians);
+  double sideLength2 = 2 * (barrelLayerRMid + detectorVolumeThickness/2.0) * std::tan(shapeAngle_radians);  
```
can you add a comment here, saying they are the two sides of the trapezoid?

BrieucF commented on FCCee/IDEA/compact/IDEA_o1_v03/MuonSystem_o1_v01.xml:
```diff
@@ -0,0 +1,70 @@
+  <readouts>
+    <readout name="MuonSystemCollection">
+     <segmentation type="CartesianGridYZ" grid_size_y="1.2*mm" grid_size_z="1.2*mm"/>   <!-- Depending on strip pitch 1.4 mm  -->
+     <id>system:1,type:2,layer:4,chamber:15,slice:1,y:-10,z:-10</id> <!-- The bit field is divided into 2^5 systems(IDEA sub-detectors), 2^4 layers(Muon System layers"barrel and endcap layers"), 2^11 chambers(the number of muRWELL chambers in every layer), 2^1 slice(number of sensitive layers inside every chambers), and 2^10 y&z strips in every sensitive layer-->
```
```suggestion
     <id>system:5,type:2,layer:4,chamber:15,slice:1,y:-10,z:-10</id> <!-- The bit field is divided into 2^5 systems(IDEA sub-detectors), 2^4 layers(Muon System layers"barrel and endcap layers"), 2^11 chambers(the number of muRWELL chambers in every layer), 2^1 slice(number of sensitive layers inside every chambers), and 2^10 y&z strips in every sensitive layer-->
```

BrieucF commented on FCCee/IDEA/compact/README.md:
```diff
@@ -9,7 +9,7 @@ IDEA version picked from the latest version in FCCDetectors repo
-Based on o1_v01 but with a detailed description of the vertex detector and the drift chamber, a place holder solenoid and the endplate absorber. Missing: SiWrapper, calorimeter, detailed muons system (coming soon).
+Based on o1_v01 but with a detailed description of the vertex detector, drift chamber and muon system, a place holder solenoid and the endplate absorber. Missing: SiWrapper, calorimeter.
 
 IDEA_o1_v03
```
I propose to leave IDEA_o1_v02 untouched in this PR and to adapt the IDEA_o1_v03 entry in the README

BrieucF commented on detector/muonSystem/README.md:
```diff
@@ -0,0 +1,10 @@
+## muonSystemMuRWELL_o1_v01.cpp
+The first version of the detailed muon system driver, it can be used to describe both barrel and endcap, if you want to eleminate one of them, just set the number of layers= 0.
+The code has been designed to be very flexible, where the user can choose the number of sides in the R-Phi plane, `numSides` (hexagon, octagon, etc), and the detector builder will automatically calculate the number and places of the copied chambers. Some of the code advantages: 
+ * If the side length do not fit with an integer number of 50 × 50 cm² , the builder can make a chamber with unusual dimensions, which can fit the excess area at the end of the side.
```
can make or will make?

BrieucF commented on FCCee/IDEA/compact/IDEA_o1_v03/MuonSystem_o1_v01.xml:
```diff
@@ -0,0 +1,70 @@
+      <dimensions x="mRWELLTotalThickness/2.0" y="mRWELLYLength" z="mRWELLZLength" x_offset="0*mm" y_offset="0*mm" z_offset="0*mm" material="Air"/>
+      <sensitive type="tracker"/>
+
+      <!-- Specify the detector parameters and the overlap /// if you want exclude any component, e.g: endcap, just put endcapDetectorParameters=0   // radius is put in the middle, so its not the inner neither the outer -->
```
```suggestion
      <!-- Specify the detector parameters and the overlap /// if you want exclude any component, e.g: endcap, just put numDetectorLayers=0 in the Endcap entry  // radius is put in the middle, so its not the inner neither the outer -->
```
?

mahmoudali2 commented:
Thanks @BrieucF and @andresailer for your comments,
I have refactored all the builder, and here are the main updates:

- Looping over layers, which led to having all the muon-system layers represented by just  one detector in the xml.
- Decreasing the number of parameters inserted by the user --> that made the code more automized and made the xml smaller and simpler.
- Refactoring the bit-field of the muon-system, and having all the muon hits in one collection called `MuonSystemcollection`, and you can distinguish between the barrel hits and the endcap hits by their different bit field section `type` if `type = 0` --> Barrel Hits, if `type = 1` --> Positive Endcap Hits, and `type = -1` --> Negative Endcap Hits.
- The ability to make multiple layers and still the last barrel layer will enclose all the endcap layers inside.
Let me know if you have any more comments.

BrieucF commented on detector/muonSystem/README.md:
```diff
@@ -0,0 +1,10 @@
+## muonSystemMuRWELL_o1_v01.cpp
+The first version of the detailed muon system driver, it can be used to describe both barrel and endcap, if you want to eleminate one of them, just set the number of layers= 0.
+The code has been designed to be very flexible, where the user can choose the number of sides in the R-Phi plane, `numSides` (hexagon, octagon, etc), and the detector builder will automatically calculate the number and places of the copied chambers. Some of the code advantages: 
+ * If the side length do not fit with an integer number of 50 × 50 cm² , the builderwill make a chamber with unusual dimensions, which can fit the excess area at the end of the side.
```
```suggestion
 * If the side length do not fit with an integer number of 50 × 50 cm² , the builder will make a chamber with unusual dimensions, which can fit the excess area at the end of the side.
```

BrieucF commented on FCCee/IDEA/compact/README.md:
```diff
@@ -9,10 +9,10 @@ IDEA version picked from the latest version in FCCDetectors repo
 ------------
 
-Based on o1_v02 but replacing the drift chamber (o1, v01) for the lightweight implementation based on twisted tubes (o1, v02). Production threshold and step limit physics have to be tuned
+Based on o1_v02 but replacing the drift chamber (o1, v01) for the lightweight implementation based on twisted tubes (o1, v02), and Add a detailed version of the muon system. Production threshold and step limit physics have to be tuned
```
```suggestion
Based on o1_v02 but replacing the drift chamber (o1, v01) for the lightweight implementation based on twisted tubes (o1, v02). NB: production threshold and step limit physics have to be tuned for the drift chamber. July 2024: Added a detailed version of the muon system. 
```

BrieucF commented on FCCee/IDEA/compact/IDEA_o1_v03/MuonSystem_o1_v01.xml:
```diff
@@ -0,0 +1,69 @@
+  <readouts>
+    <readout name="MuonSystemCollection">
+     <segmentation type="CartesianGridYZ" grid_size_y="1.2*mm" grid_size_z="1.2*mm"/>   <!-- Depending on strip pitch 1.4 mm  -->
+     <id>system:5,type:2,layer:4,chamber:15,slice:1,y:-10,z:-10</id> <!-- The bit field is divided into 2^5 systems(IDEA sub-detectors), 2^2 types(Barrel Muon System, positive endcap, and negative endcap"), 2^4 layers(number of layers in barrel for example) ,2^15 chambers(the number of muRWELL chambers in every layer), 2^1 slice(number of sensitive layers inside every chambers), and 2^10 y&z stri ...
```
```suggestion
     <id>system:5,type:2,layer:4,chamber:15,slice:1,y:-10,z:-10</id> <!-- The bit field is divided into 2^5 systems(IDEA sub-detectors), 2^2 types(0 = Barrel Muon System, +1 = positive endcap and -1 = negative endcap"), 2^4 layers(number of layers in barrel for example) ,2^15 chambers(the number of muRWELL chambers in every layer), 2^1 slice(number of sensitive layers inside every chambers), and 2^10 y&z strips in every sensitive layer-->
```?
I'd also propose to change `type` to `side`, but it is not a strong request.

mahmoudali2 replied:
I'm afraid using `side` instead of `type` would cause confusion with the detector side variable that I use to determine the polygon shape sides.

BrieucF replied:
Ok! But can you still say in the xml which value is given to barrel/endcap etc?

mahmoudali2 replied:
Yes, I just added it in the last commit

atolosadelgado commented on FCCee/IDEA/compact/IDEA_o1_v03/MuonSystem_o1_v01.xml:
```diff
@@ -0,0 +1,69 @@
+      <slice x="GasLayerThick/2.0" material="ARCO2CF4" sensitive="true" vis="Sensitive_vis" />
+      <slice x="Cu2Thick/2.0" material="G4_Cu" vis="Cu_vis" />
+      <slice x="KaptonThick/2.0" material="Kapton" vis="Kapton_vis" />
+      <slice x="CarnonFiberThick/2.0" material="CarbonFiber" vis="CarbonFiber_vis" />
```
```suggestion
      <slice x="CarbonFiberThick/2.0" material="CarbonFiber" vis="CarbonFiber_vis" />
```

atolosadelgado commented on FCCee/IDEA/compact/IDEA_o1_v03/materials_o1_v02.xml:
```diff
@@ -395,4 +395,207 @@
+    <composite n="0.40" ref="CF4"/>
+  </material>
+
+  <material name="G4_Fe" state="solid">
```
Are these materials still needed?

mahmoudali2 replied:
Yes, it's the material of the uRWELL.

atolosadelgado commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,946 @@
+  dd4hep::Transform3D unionTransform2(unionRot, unionPos2);
+
+  //Combining two shapes by UnionSolid: the first shape is centralized and the second transform around the first..
+  dd4hep::UnionSolid barrelAndPositiveEndcap(BarrelEnv, EndcapEnv, unionTransform);
```
maybe a cylinder would be more performant?

atolosadelgado commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,946 @@
+  dd4hep::Rotation3D barrelUnionRot(dd4hep::RotationY(0.0 * dd4hep::degree));
+  dd4hep::Transform3D barrelUnionTransform(barrelUnionRot, barrelUnionPos);
+
+  dd4hep::UnionSolid barrelUnion(BarrelEnvWithoutLastLayer, BarrelLastLayerEnv, barrelUnionTransform);
```
maybe a cylinder would be more performant?

atolosadelgado commented on detector/muonSystem/muonSystemMuRWELL_o1_v01.cpp:
```diff
@@ -0,0 +1,946 @@
+      dd4hep::Volume sideVol2(sideName, sideEnvelope2, mat);      
+
+      double angle_degrees = shapeAngle * side; // Calculate the angle for each side
+      double angle_radians = angle_degrees * M_PI / 180.0;
```
to avoid confusion I would suggest to work on angle natural units of dd4hep. In addition, conversion of deg <-> rad can be done using ROOT factos`TMath::DegToRad()`

atolosadelgado reviewed (commented):
Overall it is good, here are minimal comments, @BrieucF can tell if they can be ignored :)

* typo: CarnonFiber :)
* Is there a real need of defining the following materials? G4_Fe, G4_Si0x56335d3d1480, Ca400x5633439c2580, ... G10_FR40x5633439ca570. These materials are introduced in this PR, lines 490-645 of the material XML file
 https://github.com/key4hep/k4geo/blob/931d8d1e96ac5f05e8239db4c2319207f11741f4/FCCee/IDEA/compact/IDEA_o1_v02/materials_o1_v01.xml

* would it be possible to the replace the union solids that make up the envelopes by cylinders for example?

BrieucF commented:
I'd say the baseline is to implement them all :-P if some are causing problems or require significant change we de-discuss ;-)

mahmoudali2 commented:
Hi @atolosadelgado , Thanks very much for your comments, here are some clarifications;

- For the materials you mentioned in IDEA V02, it's now removed. I have removed all updates on muon-system from the V02, and just added it to V03. Also no more hex-names, but G4_Fe is needed for the yokes.
- Replacing union solids with cylinders will cause overlaps between the muon-system and other sub-detectors especially the DR-Calo. the idea of using union solid is to combine the different layers mother volumes together since there is no specific shapes that can describe the MS-barrel which is combination of two different polyhedra with different z-lengths(photo attached).
<img width="1001" alt="Screenshot 2024-07-05 at 17 41 22" src="https://github.com/key4hep/k4geo/assets/120766336/5fb5d442-edb6-46d5-b370-46f1bc55cfa5">
The same case for all the muon-system, which is 3 combinations of 3 different polyhedra: 
<img width="774" alt="Screenshot 2024-07-05 at 17 44 25" src="https://github.com/key4hep/k4geo/assets/120766336/7bc53d38-5bf6-48e2-b6f9-a5a8af0f2bbf">

BrieucF commented on FCCee/IDEA/compact/IDEA_o1_v03/MuonSystem_o1_v01.xml:
```diff
@@ -0,0 +1,69 @@
+       <constant name = "CarbonFiberThick"                      value = "0.0001*mm"/>
+       <constant name = "CarbonFiber2Thick"                     value = "0.1*mm"/>
+       <constant name = "SiThick"                               value = "1.6*mm"/>
+       <constant name = "mRWELLTotalThickness"                  value = "G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick+KaptonThick+CarnonFiberThick+CuThick+CarbonFiber2Thick+CuThick+SiThick"/> <!-- This sequense is the current order of mRWELL (total 10 slices) -->
```
Tests are crashing, forgot one ;-)

```suggestion
       <constant name = "mRWELLTotalThickness"                  value = "G10_FR4Thick+CuThick+GasLayerThick+Cu2Thick+KaptonThick+CarbonFiberThick+CuThick+CarbonFiber2Thick+CuThick+SiThick"/> <!-- This sequense is the current order of mRWELL (total 10 slices) -->
```

atolosadelgado commented:
> Hi @atolosadelgado , Thanks very much for your comments, here are some clarifications;
> 
>     * For the materials you mentioned in IDEA V02, it's now removed. I have removed all updates on muon-system from the V02, and just added it to V03. Also no more hex-names, but G4_Fe is needed for the yokes.
> 
>     * Replacing union solids with cylinders will cause overlaps between the muon-system and other sub-detectors especially the DR-Calo. the idea of using union solid is to combine the different layers mother volumes together since there is no specific shapes that can describe the MS-barrel which is combination of two different polyhedra with different z-lengths(photo attached).
> 
> 
> <img alt="Screenshot 2024-07-05 at 17 41 22" width="1001" src="https://private-user-images.githubusercontent.com/120766336/346613248-5fb5d442-edb6-46d5-b370-46f1bc55cfa5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0NTIzNjIsIm5iZiI6MTcyMDQ1MjA2MiwicGF0aCI6Ii8xMjA3NjYzMzYvMzQ2NjEzMjQ4LTVmYjVkNDQyLWVkYjYtNDZkNS1iMzcwLTQ2ZjFiYzU1Y2ZhNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwOFQxNTIxMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MzFjZDE2ZGYxZDUwZGRjY2M5NTkxNTkzZGM3Mzg5NGRkYWRkZjUzOTQ4YzhkNjRjMzgzZWQyM2VlZGU4YTQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.iLr5eitfczzyC62dZNUSVCHsKqWc74oprldJcPZW-kk"> The same case for all the muon-system, which is 3 combinations of 3 different polyhedra: <img alt="Screenshot 2024-07-05 at 17 44 25" width="774" src="https://private-user-images.githubusercontent.com/120766336/346613528-7bc53d38-5bf6-48e2-b6f9-a5a8af0f2bbf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0NTIzNjIsIm5iZiI6MTcyMDQ1MjA2MiwicGF0aCI6Ii8xMjA3NjYzMzYvMzQ2NjEzNTI4LTdiYzUzZDM4LTViZjYtNDhlMi1iNmY5LWE1YThhZjBmMmJiZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwOFQxNTIxMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWQ2MjVlMDY4OTY5NzU0Y2NkOTk4MGExZDk3MjllNTBiNjI2YWNmMWI2OWIyMzExY2IwNWVmMDQ5YTk1MTE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bPI79pKLULBl9YyeXvvSsQo2PSt7nehWbD1epEx1sFk">

I had a chat with Brieuc, and these are minor comments which can be ignored
